### PR TITLE
Add collector for the launchd table

### LINKED ISSUE #31429: Add collector for the launchd table
| Wazuh version | Install type | Install method     | Platform |
|---------------|--------------|--------------------|----------|
| 4.14.0        | Agent        | Packages / Sources | macOS    |

# Description

This issue proposes the implementation of a new **macOS-only collector** for the `services` inventory table within the `syscollector` module.  
The collector will use the `extended_sources` submodule to retrieve metadata from **launchd jobs** (daemons and agents defined via `.plist` files).  

The collected information will be normalized to the inventory schema, aligned with ECS-compatible fields where possible, and integrated into the stateful data model.  
This will enhance observability of macOS environments, providing visibility into service configuration, execution state, startup behavior, and security-relevant settings.

**Relevant fields:**

| Column              | Type     | Description                                                                 |
|---------------------|----------|-----------------------------------------------------------------------------|
| `path`              | text     | Path to daemon or agent plist                                               |
| `name`              | text     | File name of the plist                                                      |
| `label`             | text     | Daemon or agent service name (launchd label)                                |
| `program`           | text     | Path to target program                                                      |
| `program_arguments` | text     | Command-line arguments passed to the program                                |
| `run_at_load`       | text     | Whether the job runs at load                                                |
| `keep_alive`        | text     | Whether the process should be restarted if killed                           |
| `on_demand`         | text     | Deprecated key, replaced by `keep_alive`                                    |
| `disabled`          | text     | Skip loading this daemon or agent on boot                                   |
| `username`          | text     | User account under which the daemon or agent runs                           |
| `groupname`         | text     | Group account under which the daemon or agent runs                          |
| `stdout_path`       | text     | Path for redirected standard output                                         |
| `stderr_path`       | text     | Path for redirected standard error                                          |
| `start_interval`    | text     | Frequency (in seconds) to run                                               |
| `watch_paths`       | text     | Launch when the specified path is modified                                  |
| `queue_directories` | text     | Launch when the specified directory is non-empty                            |
| `inetd_compatibility` | text   | Run as if launched from inetd                                               |
| `start_on_mount`    | text     | Run every time a filesystem is mounted                                      |
| `root_directory`    | text     | Directory to chroot to before launch                                        |
| `working_directory` | text     | Directory to chdir to before launch                                         |
| `process_type`      | text     | Intended purpose/type of the job                                            |

# Tasks

- [x] Implement a new `LaunchdServiceProvider` in `extended_sources`  
- [x] Map launchd fields into ECS-compatible and custom inventory fields  
- [x] Normalize booleans (`disabled`, `keep_alive`, `run_at_load`) to `true/false`  
- [x] Add unit tests for parsing `.plist` and handling special cases (arrays, dictionaries)  
- [x] Add integration tests across supported macOS versions  

# Acceptance Criteria

- Launchd job metadata is successfully retrieved and structured in JSON format  
- Collector works across supported macOS versions (Big Sur, Monterey, Ventura, Sequoia)  
- Runtime information (`pid`, `state`) is collected using `launchctl list` when available  
- Extended fields (`watch_paths`, `queue_directories`) are stored in JSON under `service.ext.launchd.*`  
- Tests confirm correct behavior, coverage of edge cases, and delta generation  

# References

- [osquery launchd table](https://github.com/osquery/osquery/blob/master/specs/darwin/launchd.table)  
- [Apple launchd.plist documentation](https://www.launchd.info/)

### PR DESCRIPTION
**Pull request by cborla**
Closes #31429

## Description

Implements a new **macOS-only collector** for the `services` inventory table within the `syscollector` module.  
The collector will use the `extended_sources` submodule to retrieve metadata from **launchd jobs** (daemons and agents defined via `.plist` files).  

The information collected will be normalized to the inventory schema, aligned with ECS-compatible fields where possible, and integrated into the stateful data model.  
This will enhance observability of macOS environments, providing visibility into service configuration, execution state, startup behavior, and security-relevant settings.

**Relevant fields:**

| Column              | Type     | Description                                                                 |
|---------------------|----------|-----------------------------------------------------------------------------|
| `path`              | text     | Path to daemon or agent plist                                               |
| `name`              | text     | File name of the plist                                                      |
| `label`             | text     | Daemon or agent service name (launchd label)                                |
| `program`           | text     | Path to target program                                                      |
| `program_arguments` | text     | Command-line arguments passed to the program                                |
| `run_at_load`       | text     | Whether the job runs at load                                                |
| `keep_alive`        | text     | Whether the process should be restarted if killed                           |
| `on_demand`         | text     | Deprecated key, replaced by `keep_alive`                                    |
| `disabled`          | text     | Skip loading this daemon or agent on boot                                   |
| `username`          | text     | User account under which the daemon or agent runs                           |
| `groupname`         | text     | Group account under which the daemon or agent runs                          |
| `stdout_path`       | text     | Path for redirected standard output                                         |
| `stderr_path`       | text     | Path for redirected standard error                                          |
| `start_interval`    | text     | Frequency (in seconds) to run                                               |
| `watch_paths`       | text     | Launch when the specified path is modified                                  |
| `queue_directories` | text     | Launch when the specified directory is non-empty                            |
| `inetd_compatibility` | text   | Run as if launched from inetd                                               |
| `start_on_mount`    | text     | Run every time a filesystem is mounted                                      |
| `root_directory`    | text     | Directory to chroot to before launch                                        |
| `working_directory` | text     | Directory to chdir to before launch                                         |
| `process_type`      | text     | Intended purpose/type of the job                                            |

## Proposed Changes

- [ ] Implement a new `LaunchdServiceProvider` in `extended_sources`   
- [ ] Normalize booleans (`disabled`, `keep_alive`, `run_at_load`) to `true/false`  
- [ ] Add unit tests for parsing `.plist` and handling special cases (arrays, dictionaries)  
- [ ] Add integration tests across supported macOS versions  

### Results and Evidence

<details><summary>Source item example</summary>

```json
    {
        "disabled": "",
        "groupname": "_networkd",
        "inetd_compatibility": "",
        "keep_alive": "",
        "label": "com.apple.symptomsd",
        "name": "com.apple.symptomsd.plist",
        "on_demand": "",
        "path": "/System/Library/LaunchDaemons/com.apple.symptomsd.plist",
        "process_type": "",
        "program": "/usr/libexec/symptomsd",
        "program_arguments": "/usr/libexec/symptomsd",
        "queue_directories": "",
        "root_directory": "",
        "run_at_load": "",
        "start_interval": "",
        "start_on_mount": "",
        "stderr_path": "",
        "stdout_path": "",
        "username": "_networkd",
        "watch_paths": "/Library/Preferences/com.apple.symptoms.plist",
        "working_directory": ""
    }
```

</details>

<details><summary>lunchd collector item example</summary>

```json
    {
        "disabled": "",
        "groupname": "_networkd",
        "inetd_compatibility": "",
        "keep_alive": "",
        "label": "com.apple.symptomsd",
        "name": "com.apple.symptomsd.plist",
        "on_demand": "",
        "path": "/System/Library/LaunchDaemons/com.apple.symptomsd.plist",
        "process_type": "",
        "program": "/usr/libexec/symptomsd",
        "program_arguments": "/usr/libexec/symptomsd",
        "queue_directories": "",
        "root_directory": "",
        "run_at_load": "",
        "start_interval": "",
        "start_on_mount": "",
        "stderr_path": "",
        "stdout_path": "",
        "username": "_networkd",
        "watch_paths": "/Library/Preferences/com.apple.symptoms.plist",
        "working_directory": ""
    }
```

</details> 

#### macOS Agent Packages:
- [intel64](https://github.com/wazuh/wazuh-agent-packages/actions/runs/17080060641)
- [arm64](https://github.com/wazuh/wazuh-agent-packages/actions/runs/17080065166)

### Artifacts Affected

- `syscollector` module (Darwin)
- `extended_sources` submodule
- `Services` inventory table
- Unit test suites for macOS

### Configuration Changes

- NA

### Tests Introduced

- Unit test suites for macOS

## Review Checklist

- [x] Code changes reviewed
- [x] Relevant evidence provided
- [x] Tests cover the new functionality
- [ ] Configuration changes documented
- [ ] Developer documentation reflects the changes
- [x] Meets requirements and/or definition of done
- [x] No unresolved dependencies with other issues

### CODE REVIEW

File: src/data_provider/src/extended_sources/services/include/launchd_darwin.hpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        /// @param path Path to the plist file.
+        /// @param service LaunchdService structure to fill with parsed data.
+        /// @return True if parsing was successful, false otherwise.
+        bool parsePlistFile(const std::string& path, LaunchdService& service);
+
+        /// @brief Joins array elements into a space-separated string.
+        /// @param arrayElements Vector of strings to join.
+        /// @return Joined string.
+        std::string joinArrayElements(const std::vector<std::string>& arrayElements);
+};
```
Comment (cborla) [inline src/data_provider/src/extended_sources/services/include/launchd_darwin.hpp]: add a new line.

File: src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+            CFTypeRef value = CFDictionaryGetValue(dict, key);
+
+            if (value)
+            {
+                if (CFGetTypeID(value) == CFStringGetTypeID())
+                {
+                    CFStringRef stringValue = static_cast<CFStringRef>(value);
+                    CFIndex length = CFStringGetLength(stringValue);
+                    CFIndex maxSize = CFStringGetMaximumSizeForEncoding(length, kCFStringEncodingUTF8) + 1;
+                    std::vector<char> buffer(maxSize);
```
Comment (cborla) [inline src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp:239]: What happens if maxSize returns an invalid or negative value? It would be useful to validate whether maxSize > 0 before define teh vector.

Comment (nbertoldo) [reply src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp:239]: Done

File: src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    if (CFGetTypeID(plist) != CFDictionaryGetTypeID())
+    {
+        CFRelease(plist);
+        return false;
+    }
+
+    CFDictionaryRef dict = static_cast<CFDictionaryRef>(plist);
+
```
Comment (cborla) [inline src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp:214]: I think we should release CFRelease(plist);, maybe in this line.

File: src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            }
+        }
+    }
+
+    // Search user-specific paths
+    // Get all home directories
+    setpwent();
+    struct passwd* pw;
+
+    while ((pw = getpwent()) != nullptr)
```
Comment (cborla) [inline src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp]: Here we could analyze what happens if we have an exception from:
- std::string homeDir(pw->pw_dir); 
- launchers.push_back(fullPath); 

Can we wrap the while loop in a try catch? to ensure the endpwent();
```suggestion
    setpwent();
    try {
        struct passwd* pw;
        while ((pw = getpwent()) != nullptr)
        {
        .
        .
        .
    }
    catch (...)
    {
        // ensure cleanup
        endpwent();
        throw; // Re-run the exeption
    }
    endpwent(); 
```

File: src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp (hunk 3 above)
Comment (nbertoldo) [reply src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp:214]: It is being released at the end. It is not correct to release it earlier because it is being used.
https://github.com/wazuh/wazuh/blob/73b933d19924fa8e61cdff29bfd0c744ad7da70d/src/data_provider/src/extended_sources/services/src/launchd_darwin.cpp#L334

**Comment by cborla** [comment]
LGTM!
- [x] Code changes reviewed
- [x] Relevant evidence provided
- [x] Tests cover the new functionality
- [ ] Configuration changes documented
- [ ] Developer documentation reflects the changes
- [x] Meets requirements and/or definition of done
- [x] No unresolved dependencies with other issues

**Review by nbertoldo — APPROVED** [review]
LGTM

- [x] Code changes reviewed
- [x] Relevant evidence provided
- [x] Tests cover the new functionality
- [ ] Configuration changes documented
- [ ] Developer documentation reflects the changes
- [x] Meets requirements and/or definition of done
- [x] No unresolved dependencies with other issues
